src/prof_recent.c

```diff
@@ -0,0 +1,470 @@
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	return prof_recent_alloc_max_load_no_lock();
+}
+
+static inline ssize_t
+prof_recent_alloc_max_exchange(tsd_t *tsd, ssize_t max) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	return atomic_exchange_zd(&prof_recent_alloc_max, max, ATOMIC_RELAXED);
+}
+
+static inline void
+increment_recent_count(tsd_t *tsd, prof_tctx_t *tctx) {
+	++tctx->recent_count;
+}
+
+inline bool
+prof_recent_alloc_prepare(tsd_t *tsd, prof_tctx_t *tctx) {
+	assert(opt_prof && prof_booted);
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), tctx->tdata->lock);
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	/*
+	 * Check whether last-N mode is turned on without trying to acquire the
+	 * lock, so as to optimize for the following two scenarios:
+	 * (1) Last-N mode is switched off;
+	 * (2) Dumping, during which last-N mode is temporarily turned off so
+	 *     as not to block sampled allocations.
+	 */
+	if (prof_recent_alloc_max_load_no_lock() == 0) {
+		return false;
+	}
+
+	/*
+	 * Increment recent_count to hold the tctx so that it won't be gone
+	 * even after tctx->tdata->lock is released.  This acts as a
+	 * "placeholder"; the real recording of the allocation requires a lock
+	 * on prof_recent_alloc_mtx and is done in prof_recent_alloc (when
+	 * tctx->tdata->lock has been released).
+	 */
+	increment_recent_count(tsd, tctx);
+	return true;
+}
+
+static void
+decrement_recent_count(tsd_t *tsd, prof_tctx_t *tctx) {
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	assert(tctx != NULL);
+	malloc_mutex_lock(tsd_tsdn(tsd), tctx->tdata->lock);
+	assert(tctx->recent_count > 0);
+	--tctx->recent_count;
+	prof_tctx_try_destroy(tsd, tctx);
+}
+
+static inline prof_recent_t *
+extent_prof_recent_alloc_get_no_lock(const extent_t *extent) {
+	return (prof_recent_t *)atomic_load_p(
+	    &extent->e_prof_info.e_prof_recent_alloc, ATOMIC_RELAXED);
+}
+
+STATIC_IF_NOT_TEST inline prof_recent_t *
+extent_prof_recent_alloc_get(tsd_t *tsd, const extent_t *extent) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	prof_recent_t *recent_alloc =
+	    extent_prof_recent_alloc_get_no_lock(extent);
+	assert(recent_alloc == NULL || recent_alloc->alloc_extent == extent);
+	return recent_alloc;
+}
+
+static inline prof_recent_t *
+extent_prof_recent_alloc_exchange(tsd_t *tsd, extent_t *extent,
+    prof_recent_t *recent_alloc) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	return (prof_recent_t *)atomic_exchange_p(
+	    &extent->e_prof_info.e_prof_recent_alloc, recent_alloc,
+	    ATOMIC_RELAXED);
+}
+
+static inline void
+extent_prof_recent_alloc_set(tsd_t *tsd, extent_t *extent,
+    prof_recent_t *recent_alloc) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	assert(recent_alloc != NULL);
+	assert(extent_prof_recent_alloc_exchange(tsd, extent, recent_alloc) ==
```

Comment:
Never do updating operations in assertions. They won't be included in non dbg builds.